### PR TITLE
exposed rebinned output to python and fix load save as nxs (no ticket)

### PR DESCRIPTION
**Pull request by geishm-ansto**
Changes to the finalize/unfinalize method for RebinnedOutput to correct inconsistent behaviour, exposed to the methods to python and fixed load and saving rebinned output workspaces as nexus files.

The changes are covered in the modifed python test scripts. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by Matthew-Andrew — COMMENTED** [review]
Code changes look good and tidy up how fractional rebinning is handled nicely. One query below.

File: Framework/Algorithms/src/SumSpectra.cpp [hunk 1]
```diff
@@ -512,14 +516,17 @@ void SumSpectra::doFractionalSum(MatrixWorkspace_sptr outputWorkspace,
   // First, we need to clean the input workspace for nan's and inf's in order
   // to treat the data correctly later. This will create a new private
   // workspace that will be retrieved as mutable.
-  auto localworkspace = replaceSpecialValues();
+  auto localworkspace = replaceSpecialValues(true);
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/Algorithms/src/SumSpectra.cpp]: Previously it didn't modify the input workspace but now it unfinalizes() the input before it completes the sum. The exec() call treats the input as a const so I made a copy.

**Comment by geishm-ansto** [comment]
Hi Mat, if you are not happy with the additional copy I can skip the unfinalize call and restore the sumspectra but also check if the the input is normalized and if it uses squared errors and apply the scale accordingly. However, it will still make a copy if the replace special values is true for the algorithm.

**Comment by geishm-ansto** [comment]
@Mantid-Matthew Hi Mat can we try to close this off soon? What would you like?

**Comment by Mantid-Matthew** [comment]
I think I would prefer if a copy was not made unless absolutely necessary. I @martyngigg also had some comments to make on the PR when I discussed it with him.

File: Framework/Algorithms/src/Rebin2D.cpp [hunk 2]
```diff
@@ -155,7 +155,7 @@ void Rebin2D::exec() {
   }
   PARALLEL_CHECK_INTERUPT_REGION
   if (useFractionalArea) {
-    outputRB->finalize(true, true);
+    outputRB->finalize(true);
```
Comment (martyngigg) [inline Framework/Algorithms/src/Rebin2D.cpp]: Can I ask why the force argument is now removed? #20953 had a fairly detailed look at this and this change was added there so I'd like to confirm we haven't broken what was done there.

File: Framework/Algorithms/src/SumSpectra.cpp [hunk 3]
```diff
@@ -549,20 +556,18 @@ void SumSpectra::doFractionalSum(MatrixWorkspace_sptr outputWorkspace,
       for (size_t yIndex = 0; yIndex < m_yLength; ++yIndex) {
         const double yErrorsVal = YErrors[yIndex];
         if (std::isnormal(yErrorsVal)) { // is non-zero, nan, or infinity
-          const double errsq =
-              yErrorsVal * yErrorsVal * FracArea[yIndex] * FracArea[yIndex];
+          const double errsq = yErrorsVal * yErrorsVal;
           YErrorSum[yIndex] += errsq;
           Weight[yIndex] += 1. / errsq;
-          YSum[yIndex] += YValues[yIndex] * FracArea[yIndex] / errsq;
```
Comment (martyngigg) [inline Framework/Algorithms/src/SumSpectra.cpp]: Can I ask why this is not correct and how it differs from doing the unfinalize and simple sum?

File: Framework/Algorithms/src/SumSpectra.cpp [hunk 4]
```diff
@@ -159,6 +159,10 @@ void SumSpectra::init() {
                   "If enabled floating point special values such as NaN or Inf"
                   " are removed before the spectra are summed.");
 
+  declareProperty("UseFractionalArea", true,
```
Comment (martyngigg) [inline Framework/Algorithms/src/SumSpectra.cpp]: We need to be careful when adding new arguments as they can affect the order in the automatically-generated Python function. In this case if a user already has a call to `SumSpectra` with all positional arguments inserting the new argument here will change the meaning of their existing call. Can we add this after `MultiplyBySpectra` ?

File: Framework/Algorithms/src/Rebin2D.cpp (hunk 2 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/Rebin2D.cpp]: @Mantid-Matthew The one case I missed is if the input is a rebinned output and finalized. A problem is that rebinToFractionalOutput writes to the output workspace without updating the m_finalized flag to False. Rebin2D addressed this by forcing the finalize while ReflectometryTransform does not. I am happy to restore the force option as this is the only place that it is used although my preference is for rebinToFractionalOutput to leave the output in the correct state.

File: Framework/Algorithms/src/SumSpectra.cpp (hunk 3 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/SumSpectra.cpp]: @Mantid-Matthew Previously SumSpectra did not check if the input was finalized so if it was already unfinalized the processing was not consistent. The unfinalize call leaves the data in a known state.

File: Framework/Algorithms/src/SumSpectra.cpp (hunk 4 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/SumSpectra.cpp]: @Mantid-Matthew Good point I'll push it to the end. Thanks.

File: Framework/Algorithms/src/Rebin2D.cpp (hunk 2 above)
Comment (martyngigg) [reply Framework/Algorithms/src/Rebin2D.cpp]: @geishm-ansto Thanks for the explanation. I see what has happened. I agree that it it is better for  `rebinToFractionalOutput` to set the flag correctly. As long as the error calculation for the finalized rebinned output matches that of before then I think this is fine. Is there a test in SumSpectraTest that covers this? If not could we add one?

File: Framework/Algorithms/src/SumSpectra.cpp (hunk 3 above)
Comment (martyngigg) [reply Framework/Algorithms/src/SumSpectra.cpp]: Thanks for the explanation. I hadn't considered the case where it was unfinalized first. Thanks.

File: Framework/Algorithms/src/SumSpectra.cpp (hunk 4 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/SumSpectra.cpp]: Moved to the end.

File: Framework/Algorithms/src/SumSpectra.cpp (hunk 3 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/SumSpectra.cpp]: Added a test to SumSpectraTest to cover this case.

File: Framework/Algorithms/src/Rebin2D.cpp (hunk 2 above)
Comment (geishm-ansto) [reply Framework/Algorithms/src/Rebin2D.cpp]: The soln is not ideal as rebinToFractionalOutput is called inside a loop over (i,j) so I added a finalizeFractionalRebin that is called at the end of the loop so the flag is set once. It works.

**Comment by geishm-ansto** [comment]
@Mantid-Matthew @martyngigg I removed the workspace copy in SumSpectra by essentially going back to the original code and and adding a check if RebinnedOutput workspaces are finalized.

**Review by martyngigg — APPROVED** [review]
Thanks for making these changes @geishm-ansto. I'm happy with this now.

**Pull request merged**